Content.Server/Cargo/Systems/PricingSystem.cs

```
@@ -0,0 +1,188 @@
+            {
+                shell.WriteError($"Grid \"{i}\" doesn't exist.");
+                continue;
+            }
+
+            List<(double, EntityUid)> mostValuable = new();
+
+            var value = AppraiseGrid(mapGrid.GridEntityId, null, (uid, price) =>
+            {
+                mostValuable.Add((price, uid));
+                mostValuable.Sort((i1, i2) => i2.Item1.CompareTo(i1.Item1));
+                if (mostValuable.Count > 5)
+                    mostValuable.Pop();
+            });
+
+            shell.WriteLine($"Grid {gid} appraised to {value} credits.");
+            shell.WriteLine($"The top most valuable items were:");
+            foreach (var (price, ent) in mostValuable)
+            {
+                shell.WriteLine($"- {ToPrettyString(ent)} @ {price} credits");
+            }
+        }
+    }
+
+    private void CalculateMobPrice(EntityUid uid, MobPriceComponent component, ref PriceCalculationEvent args)
+    {
+        if (!TryComp<BodyComponent>(uid, out var body) || !TryComp<MobStateComponent>(uid, out var state))
+            throw new Exception("Tried to get the mob price of an object that isn't a mob with a body and state!");
+
+        var partList = body.Slots.ToList();
+        var totalPartsPresent = partList.Sum(x => x.Part != null ? 1 : 0);
+        var totalParts = partList.Count;
+
+        var partRatio = totalPartsPresent / (double) totalParts;
+        var partPenalty = component.Price * (1 - partRatio) * component.MissingBodyPartPenalty;
+
+        args.Price += (component.Price - partPenalty) * (state.IsAlive() ? 1.0 : component.DeathPenalty);
+    }
+
+    private void CalculateMaterialPrice(EntityUid uid, MaterialPriceComponent component, ref PriceCalculationEvent args)
+    {
+        if (!TryComp<MaterialComponent>(uid, out var material))
+            throw new Exception("Tried to get the stack price of an object that isn't a stack w/ material!");
```

does this really need to throw instead of just logging

---

should also log which entity with a toprettystring prolly